client/src/connection_cache.rs

```diff
@@ -193,88 +193,114 @@ pub fn set_use_quic(use_quic: bool) {
 fn get_connection(addr: &SocketAddr) -> (Connection, Arc<ConnectionCacheStats>) {
     let mut get_connection_measure = Measure::start("get_connection_measure");
     let mut get_connection_map_lock_measure = Measure::start("get_connection_map_lock_measure");
-    let mut map = (*CONNECTION_MAP).lock().unwrap();
-    get_connection_map_lock_measure.stop();
-
-    if map
-        .last_stats
-        .should_update(CONNECTION_STAT_SUBMISSION_INTERVAL)
-    {
-        map.stats.report();
-    }
-
-    let mut get_connection_map_measure = Measure::start("get_connection_hit_measure");
-    let (connection, hit, maybe_stats) = match map.map.get(addr) {
-        Some(connection) => {
-            let mut stats = None;
-            // update connection stats
-            if let Connection::Quic(conn) = connection {
-                stats = conn.stats().map(|s| (conn.base_stats(), s));
+    let (connection, hit, map_measure, report_stats, connection_cache_stats, maybe_stats) = {
+        let map = (*CONNECTION_MAP).read().unwrap();
+        get_connection_map_lock_measure.stop();
+
+        let report_stats = map
+            .last_stats
+            .should_update(CONNECTION_STAT_SUBMISSION_INTERVAL);
+
+        let mut get_connection_map_measure = Measure::start("get_connection_hit_measure");
+        let (connection, hit, connection_cache_stats, maybe_stats) = match map.map.get(addr) {
+            Some(connection) => {
+                let mut stats = None;
+                // update connection stats
+                if let Connection::Quic(conn) = connection {
+                    stats = conn.stats().map(|s| (conn.base_stats(), s));
+                }
+                (connection.clone(), true, map.stats.clone(), stats)
             }
-            (connection.clone(), true, stats)
-        }
-        None => {
-            let (_, send_socket) = solana_net_utils::bind_in_range(
-                IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)),
-                VALIDATOR_PORT_RANGE,
-            )
-            .unwrap();
-            let connection = if map.use_quic {
-                Connection::Quic(Arc::new(QuicTpuConnection::new(send_socket, *addr)))
-            } else {
-                Connection::Udp(Arc::new(UdpTpuConnection::new(send_socket, *addr)))
-            };
-
-            map.map.put(*addr, connection.clone());
-            (connection, false, None)
-        }
+            None => {
+                let (_, send_socket) = solana_net_utils::bind_in_range(
+                    IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)),
+                    VALIDATOR_PORT_RANGE,
+                )
+                .unwrap();
+                let connection = if map.use_quic {
+                    Connection::Quic(Arc::new(QuicTpuConnection::new(send_socket, *addr)))
+                } else {
+                    Connection::Udp(Arc::new(UdpTpuConnection::new(send_socket, *addr)))
+                };
+
+                drop(map);
```

Comment:
The moment the lock is released -- another thread could get in actually add this entry to the map. Are we concerned with that?

Reply:
Yeah, that's a good point. Let me see what can be done to fix it.

Reply:
Refactored the code a bit. It should handle the condition of when another thread could insert an entry between drop and acquiring the write lock.

Reply:
Special handling for this condition was causing the cache miss use-case to sometimes take up to 300 ms (compared to 20ms). I am removing this handling.

BTreeMap::insert() for duplicate entries should be able to handle this. The insert will remove the old entry from the map, and insert the new one. `try_insert()` would have been a better solution, as it returns an error (with current entry) if there was a duplicate insert. But that's an unstable feature.